src/message_bus.zig

```diff
@@ -18,38 +22,152 @@ pub const Message = struct {
+                    .address = address,
+                } };
+                replicas[i].connect();
+            }
+        }
 
-    const Envelope = struct {
-        address: Address,
-        message: *Message,
-        next: ?*Envelope,
-    };
+        const connections = try allocator.alloc(Connection, num_connections);
+        errdefer allocator.free(connections);
+        mem.set(Connection, connections, .{ .message_bus = self });
 
-    pub fn init(allocator: *Allocator, configuration: []ConfigurationAddress) !MessageBus {
-        var self = MessageBus{
+        self.* = .{
             .allocator = allocator,
-            .configuration = configuration,
+            .io = io,
+            .server = server,
+            .server_fd = try init_tcp(configuration[server_id]),
+            .replicas = replicas,
+            .connections = connections,
         };
-        return self;
+
+        self.maybe_accept();
     }
 
-    pub fn deinit(self: *MessageBus) void {}
+    fn init_tcp(address: std.net.Address) !os.socket_t {
+        const fd = try os.socket(address.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0);
+        errdefer os.close(fd);
+        // TODO: configure RCVBUF, SNDBUF, KEEPALIVE, TIMEOUT, NODELAY
+        try os.setsockopt(server, os.SOL_SOCKET, os.SO_REUSEADDR, &mem.toBytes(@as(c_int, 1)));
+        // TODO: port hopping
+        try os.bind(server, &address.any, address.getOsSockLen());
+        try os.listen(server, tcp_backlog);
+    }
+
+    /// Must be called whenever a Connection in the connections slice is closed.
+    fn maybe_accept(self: *MessageBus) void {
+        if (self.connections_used == self.connections.len) {
+            if (self.replica_connections == self.replicas.len - 1) {
+                // If there are no unusued connections and all replicas are
+                // connected, do nothing.
+                return;
+            } else {
+                assert(self.replica_connections < self.replicas.len - 1);
+                // There are no unused connections but not all replicas are connected.
+                // Disconnect a client to make sure there is space for a replica to connect.
+                log.info("all connections in use but not all replicas connected, disconnecting a client", .{});
+                for (self.connections) |connection| {
+                    assert(connection.peer != .none);
+                    if (connection.peer == .client) {
+                        connection.close();
+                        return;
+                    }
+                }
+                // This code should never be reached in normal circumstances.
+                // In the edge case that all client connections are currently
+                // waiting on the first message header to determine whether the peer
+                // is a client or a replica, disconnected one of these
+                // waiting connections to ensure we make progress.
```

Comment:
We should add an assertion within `init()` that `connections_max` > `configuration.len`.